apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
 
-    if (nextPanelObject.beforeEnter) {
-      promise = promise.then(
-        nextPanelObject.beforeEnter.bind(nextPanelObject, transitionArgs)
-      );
+    if (parentView === currentParentView) {
+      nextLocation = hash;
+    } else {
+      nextLocation = '/' + nextLocation + '#!' + hash;
     }
 
-    promise = promise.then(function resetPanel() {
-      // While we're sliding, we aren't in any panel
-      currentPanel = null;
-    });
+    startNavigation(
+      { newPanel: view, args, partial: true }
+    ).then(
+      () => executeNavigationStep('beforeLeave')
+    ).then(
+      () => setLocation(nextLocation)
+    ).catch(
+      onNavigationError
+    );
+    // TODO do something on pageshow?
+    //
+    return navigationTransition.defer.promise;
+  },
 
-    // sliding
-    promise = promise.then(
-      this.slide.bind(this, nextPanelInfo.wrapperPosition)
-    ).then(function changePanel() {
-      currentPanel = {
-        panel: panel,
-        args: args
-      };
-    });
+  /**
+   * This is called when a view is started.
+   *
+   * It tries to find the right view from the location, and then initiates a
+   * navigation from this.
+   */
+  init() {
+    debug('init()');
+    var newPanel = findViewFromLocation(window.location);
+    if (!newPanel) {
+      return;
+    }
 
-    // afterLeave and afterEnter should not trigger errors
-    promise = promise.then(function() {
-      var promise = Promise.resolve();
+    var args = Utils.params(decodeURIComponent(window.location.hash));
 
-      if (currentPanelObject && currentPanelObject.afterLeave) {
-        promise = promise.then(
-          currentPanelObject.afterLeave.bind(currentPanelObject, transitionArgs)
-        ).catch(catchError);
-      }
+    debug('init: found panel', newPanel.name, 'with args', args);
 
-      if (nextPanelObject.afterEnter) {
-        promise = promise.then(
-          nextPanelObject.afterEnter.bind(nextPanelObject, transitionArgs)
-        ).catch(catchError);
-      }
+    // we don't wait for the promise because we're at init and we know we have
+    // no previous state.
+    startNavigation({newPanel, args});
 
-      return promise;
-    });
+    // right away as we don't execute anything on the previous panel, and we
+    // need a state at startup.
+    setFutureState();
 
-    promise = promise.then(
-      function resolved() {
-        this.transitionPromise = null;
-        if (nextPanelContainer === currentPanelContainer) {
-          return;
-        }
-        nextPanelContainer.setAttribute('aria-hidden', false);
-        if (currentPanelContainer) {
-          currentPanelContainer.setAttribute('aria-hidden', true);
-        }
-      }.bind(this),
-      function rejected(e) {
-        catchError(e);
-        this.transitionPromise = null;
-        return Promise.reject(new Error('Error while transitioning'));
-      }.bind(this)
+    executeNavigationStep('beforeEnter').then(
+      switchPanel
```

Comment:
wondering about calling switchPanel earlier so that we see more of the view earlier